resources/tentative/todomvc/src/components/todo-item/todo-item.component.js

```diff
@@ -0,0 +1,194 @@
+                    break;
+            }
+        });
+    }
+
+    startEdit() {
+        this.item.classList.add("editing");
+        this.editInput.value = this.title;
+        this.editInput.focus();
+    }
+
+    stopEdit() {
+        this.item.classList.remove("editing");
+    }
+
+    cancelEdit() {
+        this.editInput.blur();
+    }
+
+    toggleItem() {
+        // update item first, before dispatch
+        this.setAttribute("completed", this.toggleInput.checked);
+
+        this.dispatchEvent(
+            new CustomEvent("toggle-item", {
+                detail: { id: this.id, completed: this.toggleInput.checked },
+                bubbles: true,
+            })
+        );
+    }
+
+    removeItem() {
+        // dispatch first, before updating item
+        this.dispatchEvent(
+            new CustomEvent("remove-item", {
+                detail: { id: this.id },
+                bubbles: true,
+            })
+        );
+        this.remove();
+    }
+
+    updateItem(event) {
+        if (event.target.value !== this.title) {
+            if (event.target.value.length === 0) {
+                this.removeItem();
+            } else {
+                this.setAttribute("title", event.target.value);
+                this.dispatchEvent(
+                    new CustomEvent("update-item", {
+                        detail: { id: this.id, title: event.target.value },
+                        bubbles: true,
+                    })
+                );
+            }
+        }
+
+        this.cancelEdit();
+    }
+
+    addListeners() {
+        this.toggleInput.addEventListener("change", this.toggleItem);
+        this.todoText.addEventListener(
+            "click",
+            useDoubleClick(this.startEdit, 500)
+        );
+        this.editInput.addEventListener("blur", this.stopEdit);
+        this.todoButton.addEventListener("click", this.removeItem);
+
+        this.keysListeners.forEach((listener) => listener.connect());
+    }
+
+    removeListeners() {
+        this.toggleInput.removeEventListener("change", this.toggleItem);
+        this.todoText.removeEventListener("click", this.startEdit);
+        this.editInput.removeEventListener("blur", this.stopEdit);
+        this.todoButton.removeEventListener("click", this.removeItem);
+
+        this.keysListeners.forEach((listener) => listener.disconnect());
+    }
+
+    attributeChangedCallback(property, oldValue, newValue) {
+        if (oldValue === newValue)
+            return;
+        this[property] = newValue;
+
+        if (this.isConnected)
+            this.update(property);
+
+    }
+
+    connectedCallback() {
+        this.update("id", "title", "completed");
+
+        this.keysListeners.push(
+            useKeyListener({
+                target: this.editInput,
+                event: "keyup",
+                callbacks: {
+                    ["Enter"]: this.updateItem,
```

Comment:
Not sure how this works. Did you mean to say `this.updateItem.bind(this)` or `(event) => this.updateItem(event)`?

Reply:
the event object is passed along when `this.updateItem` is called, which is already bound in the constructor.

Reply:
Oh I see.